lxd/networks_utils.go

```diff
@@ -1216,3 +1216,108 @@ func networkGetState(netIf net.Interface) api.NetworkState {
 	network.Counters = shared.NetworkGetCounters(netIf.Name)
 	return network
 }
+
+// networkGetDevMTU retrieves the current MTU setting for a named network device.
+func networkGetDevMTU(devName string) (uint64, error) {
+	content, err := ioutil.ReadFile(fmt.Sprintf("/sys/class/net/%s/mtu", devName))
+	if err != nil {
+		return 0, err
+	}
+
+	// Parse value
+	mtu, err := strconv.ParseUint(strings.TrimSpace(string(content)), 10, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	return mtu, nil
+}
+
+// networkSetDevMTU sets the MTU setting for a named network device if different from current.
+func networkSetDevMTU(devName string, mtu uint64) error {
+	curMTU, err := networkGetDevMTU(devName)
+	if err != nil {
+		return err
+	}
+
+	// Only try and change the MTU if the requested mac is different to current one.
+	if curMTU != mtu {
+		_, err := shared.RunCommand("ip", "link", "set", "dev", devName, "mtu", fmt.Sprintf("%d", mtu))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// networkGetDevMAC retrieves the current MAC setting for a named network device.
+func networkGetDevMAC(devName string) (string, error) {
+	content, err := ioutil.ReadFile(fmt.Sprintf("/sys/class/net/%s/address", devName))
+	if err != nil {
+		return "", err
+	}
+
+	return strings.TrimSpace(fmt.Sprintf("%s", content)), nil
+}
+
+// networkSetDevMAC sets the MAC setting for a named network device if different from current.
+func networkSetDevMAC(devName string, mac string) error {
+	curMac, err := networkGetDevMAC(devName)
+	if err != nil {
+		return err
+	}
+
+	// Only try and change the MAC if the requested mac is different to current one.
+	if curMac != mac {
+		_, err := shared.RunCommand("ip", "link", "set", "dev", devName, "address", mac)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// virtFuncInfo holds information about SR-IOV virtual functions.
+type virtFuncInfo struct {
+	mac        string
+	vlan       int
+	spoofcheck bool
+}
+
+// networkGetVirtFuncInfo returns info about an SR-IOV virtual function from the ip tool.
+func networkGetVirtFuncInfo(devName string, vfID int) (vf virtFuncInfo, err error) {
+	cmd := exec.Command("ip", "link", "show", devName)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return
+	}
+	if err = cmd.Start(); err != nil {
+		return
+	}
+	defer stdout.Close()
+
+	//Try and match: "vf 1 MAC 00:00:00:00:00:00, vlan 4095, spoof checking off"
+	re := regexp.MustCompile(fmt.Sprintf(`vf %d MAC ((?:[[:xdigit:]]{2}:){5}[[:xdigit:]]{2}).*, vlan (\d+), spoof checking (\w+)`, vfID))
```

Comment:
@stgraber testing on my intel based SR-IOV card this PR breaks SR-IOV support because of this regex.

This is because the output of `ip link show` is different than for the SR-IOV card on lantea.

It is is a different order, and only shows the vlan if being used, e.g.

```
18: enp4s0f1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 6c:b3:11:1c:09:53 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, vlan 1234, spoof checking on, link-state auto, trust off
    vf 1 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
    vf 2 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
    vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
    vf 4 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
    vf 5 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
    vf 6 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off
```

Do you know of a more reliable way to read the VF's properties? If not I'll update to make this function more flexible.

Reply:
@stgraber actually just looks like if VLAN is set to 0 then its not shown in output. Will update the function to understand that. On lantea the default vlan was 4095 (which I believe indicates to some switches that VLAN tagging should be extended to the host), but on Intel cards its defaulting to zero.